browser/brave_wallet/asset_discovery_manager_unittest.cc

```diff
@@ -0,0 +1,1190 @@
+        kEthereumBlockTagEarliest, kEthereumBlockTagLatest);
+    wallet_service_observer_->WaitForOnDiscoverAssetsCompleted(
+        expected_token_contract_addresses);
+    auto next_asset_discovery_from_blocks =
+        GetPrefs()->GetDict(kBraveWalletNextAssetDiscoveryFromBlocks).Clone();
+    const auto path = base::StrCat({kEthereumPrefKey, ".", chain_id});
+    const std::string* next_asset_discovery_from_block =
+        next_asset_discovery_from_blocks.FindStringByDottedPath(path);
+    if (next_asset_discovery_from_block) {
+      EXPECT_EQ(*next_asset_discovery_from_block,
+                expected_next_asset_discovery_from_block);
+    } else {
+      ASSERT_EQ(expected_next_asset_discovery_from_block, "");
+    }
+    wallet_service_observer_->Reset();
+  }
+
+  void TestDiscoverAssetsOnAllSupportedChainsAccountsAdded(
+      const std::vector<std::string>& account_addresses,
+      const std::vector<std::string>& expected_chain_ids,
+      const base::Time expected_assets_last_discovered_at_pref,
+      const base::Value::Dict expected_next_asset_discovery_from_blocks,
+      const std::vector<std::string>& expected_token_contract_addresses = {}) {
+    asset_discovery_manager_->DiscoverAssetsOnAllSupportedChainsAccountsAdded(
+        account_addresses);
+    base::RunLoop().RunUntilIdle();
+    EXPECT_EQ(GetPrefs()->GetTime(kBraveWalletLastDiscoveredAssetsAt),
+              expected_assets_last_discovered_at_pref);
+    EXPECT_EQ(GetPrefs()->GetDict(kBraveWalletNextAssetDiscoveryFromBlocks),
+              expected_next_asset_discovery_from_blocks);
+    EXPECT_FALSE(wallet_service_observer_->OnDiscoverAssetsCompletedFired());
+    wallet_service_observer_->Reset();
+  }
+
+  void TestDiscoverAssetsOnAllSupportedChainsRefresh(
+      const std::vector<std::string>& account_addresses,
+      const base::Time previous_assets_last_discovered_at,
+      base::OnceCallback<void(base::Time previous, base::Time current)>
+          assets_last_discovered_at_test_fn,
+      const base::Value::Dict previous_next_asset_discovery_from_blocks,
+      const std::vector<std::string>& expected_token_contract_addresses) {
+    asset_discovery_manager_->DiscoverAssetsOnAllSupportedChainsRefresh(
+        account_addresses);
+    wallet_service_observer_->WaitForOnDiscoverAssetsCompleted(
+        expected_token_contract_addresses);
+    base::Time current_assets_last_discovered_at =
+        GetPrefs()->GetTime(kBraveWalletLastDiscoveredAssetsAt);
+    std::move(assets_last_discovered_at_test_fn)
+        .Run(previous_assets_last_discovered_at,
+             current_assets_last_discovered_at);
+    const base::Value::Dict current_next_asset_discovery_from_blocks =
+        GetPrefs()->GetDict(kBraveWalletNextAssetDiscoveryFromBlocks).Clone();
+    EXPECT_EQ(previous_next_asset_discovery_from_blocks,
+              current_next_asset_discovery_from_blocks);
+    EXPECT_TRUE(wallet_service_observer_->OnDiscoverAssetsCompletedFired());
```

Comment:
No point for this cuz it won't quit the run_loop in WaitForOnDiscoverAssetsCompleted unless this event is fired.